epochStart/metachain/rewardsV2.go

```diff
@@ -0,0 +1,496 @@
+	accumulatedDust.Add(accumulatedDust, dust)
+
+	return aggregateBaseAndTopUpRewardsPerNode(baseRewardsPerNode, topUpRewardsPerNode), accumulatedDust
+}
+
+func (rc *rewardsCreatorV2) computeBaseRewardsPerNode(
+	validatorsInfo map[uint32][]*state.ValidatorInfo,
+	baseRewards *big.Int,
+) (map[uint32][]*big.Int, *big.Int) {
+	baseRewardsPerNode := make(map[uint32][]*big.Int)
+	accumulatedRewards := big.NewInt(0)
+
+	for shardID, valInfoList := range validatorsInfo {
+		baseRewardsPerNode[shardID] = make([]*big.Int, len(valInfoList))
+		for i, valInfo := range valInfoList {
+			baseRewardsPerNode[shardID][i] = big.NewInt(0).Mul(
+				rc.mapBaseRewardsPerBlockPerValidator[shardID],
+				big.NewInt(0).SetInt64(int64(valInfo.NumSelectedInSuccessBlocks)))
+			accumulatedRewards.Add(accumulatedRewards, baseRewardsPerNode[shardID][i])
+		}
+	}
+
+	return baseRewardsPerNode, big.NewInt(0).Sub(baseRewards, accumulatedRewards)
+}
+
+func (rc *rewardsCreatorV2) computeTopUpRewardsPerNode(
+	topUpRewards *big.Int,
+	nodesInfo map[uint32][]*state.ValidatorInfo,
+) (map[uint32][]*big.Int, *big.Int) {
+
+	topUpRewardsPerNode := make(map[uint32][]*big.Int)
+	accumulatedTopUpRewards := big.NewInt(0)
+	topUpPerNode := rc.getTopUpForAllEligibleNodes(nodesInfo)
+	topUpRewardPerShard := rc.computeTopUpRewardsPerShard(topUpRewards, topUpPerNode)
+	totalPowerInShard, nodesPower := computeNodesPowerInShard(nodesInfo, topUpPerNode)
+
+	// topUpRewardPerNodeInShardX = nodePowerInShardX*topUpRewardsShardX/totalPowerInShardX
+	for shardID, nodesPowerList := range nodesPower {
+		topUpRewardsPerNode[shardID] = make([]*big.Int, len(nodesPowerList))
+		if totalPowerInShard[shardID].Cmp(big.NewInt(0)) == 0 {
+			log.Warn("rewardsCreatorV2.computeTopUpRewardsPerNode",
+				"error", "shardPower zero",
+				"shardID", shardID)
+		}
+
+		for i, nodePower := range nodesPowerList {
+			if totalPowerInShard[shardID].Cmp(big.NewInt(0)) == 0 {
+				// avoid division by zero
+				topUpRewardsPerNode[shardID][i] = big.NewInt(0)
+				continue
+			}
+
+			topUpRewardsPerNode[shardID][i] = big.NewInt(0).Mul(nodePower, topUpRewardPerShard[shardID])
+			topUpRewardsPerNode[shardID][i].Div(topUpRewardsPerNode[shardID][i], totalPowerInShard[shardID])
+			accumulatedTopUpRewards.Add(accumulatedTopUpRewards, topUpRewardsPerNode[shardID][i])
+		}
+	}
+
+	return topUpRewardsPerNode, big.NewInt(0).Sub(topUpRewards, accumulatedTopUpRewards)
+}
+
+//      (2*k/pi)*atan(x/p), where:
+//     k is the rewards per day limit for top-up stake k = c * economics.TotalToDistribute, c - constant, e.g c = 0.25
+//     x is the cumulative top-up stake value for eligible nodes
+//     p is the cumulative eligible stake where rewards per day reach 1/2 of k (includes topUp for the eligible nodes)
+//     pi is the mathematical constant pi = 3.1415...
+func (rc *rewardsCreatorV2) computeTopUpRewards(totalToDistribute *big.Int, totalTopUpEligible *big.Int) *big.Int {
```

Comment:
we need a set of simulated data - what we expect in the mainnet and check the end result.

Reply:
will add this in a separate PR as this pull request already got too large